lib/common/ipc_attrd.c

```diff
@@ -283,234 +282,3 @@ pcmk__attrd_api_update(pcmk_ipc_api_t *api, const char *task, const char *host,
 
     return rc;
 }
-
-/*!
- * \internal
- * \brief Send an operation to pacemaker-attrd via IPC
- *
- * \param[in] ipc       Connection to pacemaker-attrd (or create one if NULL)
- * \param[in] attrd_op  XML of pacemaker-attrd operation to send
- *
- * \return Standard Pacemaker return code
- */
-static int
-send_attrd_op(crm_ipc_t *ipc, xmlNode *attrd_op)
-{
-    int rc = -ENOTCONN; // initially handled as legacy return code
-    int max = 5;
-
-    static gboolean connected = TRUE;
-    static crm_ipc_t *local_ipc = NULL;
-    static enum crm_ipc_flags flags = crm_ipc_flags_none;
-
-    if (ipc == NULL && local_ipc == NULL) {
-        local_ipc = crm_ipc_new(T_ATTRD, 0);
-        pcmk__set_ipc_flags(flags, "client", crm_ipc_client_response);
-        connected = FALSE;
-    }
-
-    if (ipc == NULL) {
-        ipc = local_ipc;
-    }
-
-    while (max > 0) {
-        if (connected == FALSE) {
-            crm_info("Connecting to cluster... %d retries remaining", max);
-            connected = crm_ipc_connect(ipc);
-        }
-
-        if (connected) {
-            rc = crm_ipc_send(ipc, attrd_op, flags, 0, NULL);
-        } else {
-            crm_perror(LOG_INFO, "Connection to cluster attribute manager failed");
-        }
-
-        if (ipc != local_ipc) {
-            break;
-
-        } else if (rc > 0) {
-            break;
-
-        } else if (rc == -EAGAIN || rc == -EALREADY) {
-            sleep(5 - max);
-            max--;
-
-        } else {
-            crm_ipc_close(ipc);
-            connected = FALSE;
-            sleep(5 - max);
-            max--;
-        }
-    }
```

Comment:
We may want to keep the try-5-times-with-sleep model for the tools that didn't pass their own IPC connection before. It's not terribly useful but it does allow tools to be called at startup.

Reply:
This could potentially live in send_attrd_request in lib/common/ipc_attrd.c,  I could check if api == NULL there and if so, do all the usual IPC setup stuff in a loop, checking to see if it connected and sleeping if not.

Reply:
That could work. We'd just have to document which functions can be used with NULL. It might be worthwhile to functionize that bit in ipc_client.c and then any IPC API function can call it where useful.

Reply:
Added a commit on top to take care of this.